Flaky test RecordedRocksDBSuite.TestPlayThenCancel

Appears to flake on master

So far this has only been confirmed to flap on mac, I do not yet have a Linux development machine to test on.

The output from the failing test is as follows:

`----------------------------------------------------------------------
FAIL: recorder_rocksdb_test.go:169: RecordedRocksDBSuite.TestPlayThenCancel

recorder_rocksdb_test.go:180:
    // Blocks until |ctx| is cancelled.
    c.Check(errors.Cause(r.player.Play(deadlineCtx, FSMHints{Log: aRecoveryLog}, r.tmpdir, bk)),
        gc.Equals, context.DeadlineExceeded)
... obtained *status.statusError = &status.statusError{Code:4, Message:"context deadline exceeded", Details:[]*any.Any(nil), XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0} ("rpc error: code = DeadlineExceeded desc = context deadline exceeded")
... expected context.deadlineExceededError = context.deadlineExceededError{} ("context deadline exceeded")
... Difference:
...     *status.statusError != context.deadlineExceededError


OOPS: 0 passed, 1 FAILED
--- FAIL: Test (0.13s)
FAIL
FAIL	github.com/LiveRamp/gazette/v2/pkg/recoverylog	1.113s`

--- a/v2/pkg/broker/teststub/broker.go
+++ b/v2/pkg/broker/teststub/broker.go
@@ -81,7 +81,7 @@ func (p *Broker) Replicate(srv pb.Journal_ReplicateServer) error {
 			} else if err != nil {
 				done = true
 
-				p.c.Check(err, gc.ErrorMatches, `rpc error: code = Canceled desc = context canceled`)
+				p.c.Check(err, gc.ErrorMatches, `rpc error: code = (Canceled|DeadlineExceeded) .*`)
 			}
 
 			log.WithFields(log.Fields{"ep": p.Endpoint(), "msg": msg, "err": err, "done": done}).Info("read")

--- a/v2/pkg/client/appender.go
+++ b/v2/pkg/client/appender.go
@@ -52,6 +52,10 @@ func (a *Appender) Write(p []byte) (n int, err error) {
 	} else {
 		n = len(p)
 	}
+
+	if err != nil {
+		err = mapGRPCCtxErr(a.ctx, err)
+	}
 	return
 }
 
@@ -61,7 +65,7 @@ func (a *Appender) Write(p []byte) (n int, err error) {
 func (a *Appender) Close() (err error) {
 	// Send an empty chunk to signal commit of previously written content
 	if err = a.lazyInit(); err != nil {
-		return
+		// Pass.
 	} else if err = a.stream.SendMsg(new(pb.AppendRequest)); err != nil {
 		// Pass.
 	} else if err = a.stream.CloseSend(); err != nil {
@@ -83,6 +87,10 @@ func (a *Appender) Close() (err error) {
 		default:
 			err = errors.New(a.Response.Status.String())
 		}
+	}
+
+	if err != nil {
+		err = mapGRPCCtxErr(a.ctx, err)
 	}
 	return
 }

--- a/v2/pkg/client/appender_test.go
+++ b/v2/pkg/client/appender_test.go
@@ -269,6 +269,29 @@ func (s *AppenderSuite) TestAppendCases(c *gc.C) {
 	c.Check(err, gc.ErrorMatches, "readerAt error")
 }
 
+func (s *AppenderSuite) TestContextErrorCases(c *gc.C) {
+	var ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+
+	var broker = teststub.NewBroker(c, ctx)
+	var rjc = pb.NewRoutedJournalClient(broker.MustClient(), pb.NoopDispatchRouter{})
+	var content = strings.NewReader("content")
+
+	// Case 1: context is cancelled.
+	var caseCtx, caseCancel = context.WithCancel(ctx)
+	caseCancel()
+
+	var _, err = Append(caseCtx, rjc, pb.AppendRequest{Journal: "a/journal"}, content)
+	c.Check(err, gc.Equals, context.Canceled)
+
+	// Case 2: context reaches deadline.
+	caseCtx, _ = context.WithTimeout(ctx, time.Microsecond)
+	<-caseCtx.Done() // Block until deadline.
+
+	_, err = Append(caseCtx, rjc, pb.AppendRequest{Journal: "a/journal"}, content)
+	c.Check(err, gc.Equals, context.DeadlineExceeded)
+}
+
 type errReaderAt struct{}
 
 func (errReaderAt) ReadAt(p []byte, off int64) (n int, err error) {

--- a/v2/pkg/client/list.go
+++ b/v2/pkg/client/list.go
@@ -66,7 +66,7 @@ func ListAll(ctx context.Context, client pb.JournalClient, req pb.ListRequest) (
 	for {
 		// List RPCs may be dispatched to any broker.
 		if r, err := client.List(pb.WithDispatchDefault(ctx), &req); err != nil {
-			return resp, err
+			return resp, mapGRPCCtxErr(ctx, err)
 		} else if err = r.Validate(); err != nil {
 			return resp, err
 		} else if r.Status != pb.Status_OK {

--- a/v2/pkg/client/list_test.go
+++ b/v2/pkg/client/list_test.go
@@ -72,6 +72,11 @@ func (s *ListSuite) TestListAllCases(c *gc.C) {
 
 	_, err = ListAll(context.Background(), broker.MustClient(), pb.ListRequest{Selector: selector})
 	c.Check(err, gc.ErrorMatches, `WRONG_ROUTE`)
+
+	// Case: It surfaces context.Canceled, rather than a gRPC error.
+	cancel()
+	_, err = ListAll(ctx, rjc, pb.ListRequest{Selector: selector, PageLimit: 10})
+	c.Check(err, gc.Equals, context.Canceled)
 }
 
 func (s *ListSuite) TestPolledList(c *gc.C) {

--- a/v2/pkg/client/reader.go
+++ b/v2/pkg/client/reader.go
@@ -11,6 +11,8 @@ import (
 
 	"github.com/LiveRamp/gazette/v2/pkg/codecs"
 	pb "github.com/LiveRamp/gazette/v2/pkg/protocol"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
 // Reader adapts a Read RPC to the io.Reader interface. It additionally supports
@@ -64,25 +66,22 @@ func (r *Reader) Read(p []byte) (n int, err error) {
 			&r.Request,
 		); err == nil {
 			n, err = r.Read(p) // Recurse to attempt read against opened |r.stream|.
-			return             // Surface read or error.
+		} else {
+			err = mapGRPCCtxErr(r.ctx, err)
 		}
+		return
 	}
 
-	if err == nil {
-		// Read and Validate the next frame.
-		if err = r.stream.RecvMsg(&r.Response); err == nil {
-			if err = r.Response.Validate(); err != nil {
-				err = pb.ExtendContext(err, "ReadResponse")
-			} else if r.Response.Status == pb.Status_OK && r.Response.Offset < r.Request.Offset {
-				err = pb.NewValidationError("invalid ReadResponse offset (%d; expected >= %d)",
-					r.Response.Offset, r.Request.Offset) // Violation of Read API contract.
-			}
-		} else if r.ctx.Err() != nil {
-			// Context cancellations are usually wrapped by augmenting errors as they
-			// return up the call stack. If our Reader's context is Done, assume
-			// that is the primary error.
-			err = r.ctx.Err()
+	// Read and Validate the next frame.
+	if err = r.stream.RecvMsg(&r.Response); err == nil {
+		if err = r.Response.Validate(); err != nil {
+			err = pb.ExtendContext(err, "ReadResponse")
+		} else if r.Response.Status == pb.Status_OK && r.Response.Offset < r.Request.Offset {
+			err = pb.NewValidationError("invalid ReadResponse offset (%d; expected >= %d)",
+				r.Response.Offset, r.Request.Offset) // Violation of Read API contract.
 		}
+	} else {
+		err = mapGRPCCtxErr(r.ctx, err)
 	}
 
 	// A note on resource leaks: an invariant of Read is that in invocations where
@@ -281,6 +280,20 @@ func InstallFileTransport(root string) (remove func()) {
 	httpClient = &http.Client{Transport: transport}
 
 	return func() { httpClient = prevClient }
+}
+
+// mapGRPCCtxErr returns ctx.Err() iff |err| represents a gRPC error with a
+// status code matching ctx.Err(). Otherwise, it returns |err| unmodified.
+// In other words, this routine "unwraps" gRPC errors which have their root cause
+// in a local Context error, instead mapping to the common context package errors.
+func mapGRPCCtxErr(ctx context.Context, err error) error {
+	if ctx.Err() == context.DeadlineExceeded && status.Code(err) == codes.DeadlineExceeded {
+		return ctx.Err()
+	}
+	if ctx.Err() == context.Canceled && status.Code(err) == codes.Canceled {
+		return ctx.Err()
+	}
+	return err
 }
 
 var (

--- a/v2/pkg/client/reader_test.go
+++ b/v2/pkg/client/reader_test.go
@@ -307,6 +307,34 @@ func (s *ReaderSuite) TestReaderSeekCases(c *gc.C) {
 	c.Check(err, gc.Equals, ErrSeekRequiresNewReader)
 }
 
+func (s *ReaderSuite) TestContextErrorCases(c *gc.C) {
+	var ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+
+	var broker = teststub.NewBroker(c, ctx)
+	var rjc = pb.NewRoutedJournalClient(broker.MustClient(), pb.NoopDispatchRouter{})
+
+	// Case 1: context is cancelled.
+	var caseCtx, caseCancel = context.WithCancel(ctx)
+	caseCancel()
+
+	var r = NewReader(caseCtx, rjc, pb.ReadRequest{Journal: "a/journal"})
+	var n, err = r.Read(nil)
+
+	c.Check(err, gc.Equals, context.Canceled)
+	c.Check(n, gc.Equals, 0)
+
+	// Case 2: context reaches deadline.
+	caseCtx, _ = context.WithTimeout(ctx, time.Microsecond)
+	<-caseCtx.Done() // Block until deadline.
+
+	r = NewReader(caseCtx, rjc, pb.ReadRequest{Journal: "a/journal"})
+	n, err = r.Read(nil)
+
+	c.Check(err, gc.Equals, context.DeadlineExceeded)
+	c.Check(n, gc.Equals, 0)
+}
+
 type readFixture struct {
 	status pb.Status
 	err    error

--- a/v2/pkg/client/retry_reader.go
+++ b/v2/pkg/client/retry_reader.go
@@ -156,7 +156,7 @@ func (rr *RetryReader) AdjustedSeek(offset int64, whence int, br *bufio.Reader) 
 
 	// Fast path: can we fulfill the seek by discarding a portion of buffered data?
 	if delta >= 0 && delta <= int64(br.Buffered()) {
-		br.Discard(int(delta))
+		_, _ = br.Discard(int(delta))
 		return offset, nil
 	}
 

--- a/v2/pkg/recoverylog/recorder_rocksdb_test.go
+++ b/v2/pkg/recoverylog/recorder_rocksdb_test.go
@@ -200,7 +200,7 @@ func (s *RecordedRocksDBSuite) TestCancelThenPlay(c *gc.C) {
 	cancelFn()
 
 	c.Check(r.player.Play(cancelCtx, FSMHints{Log: aRecoveryLog}, r.tmpdir, bk),
-		gc.ErrorMatches, `determining log head: rpc error: code = Canceled desc = context canceled`)
+		gc.ErrorMatches, `determining log head: context canceled`)
 
 	<-r.player.Done()
 }

